fritz-tools: implement fritz_cal_extract with reverse option for AVM 7430

dzzinstant opened the pull request:
Implementation of `fritz_cal_extract` with additional option to 
retrieve firmware data in reversed byte order, which is necessary
for the AVM FritzBox 7430 device. To enable reversal on the spot,
this implementation makes use of a temporary buffer.

Affected command line options:
`-r` reverse data after zlib uncompression
`-i n` skip `n` bytes of data _after_ optional reversal
`-l m` truncates resulting data to `m` bytes, after optional reversal
 and skipping of `n` bytes.
The allocated buffer will always have a minimum size of `(n + m)` bytes.

### Possibly affected devices (all that use `fritz_cal_extract`)
lantiq-xrx200: AVM FritzBox 7412, 7430
ipq-40xx: AVM FritzBox 4040, 7520, 7530; AVM FritzRepeater 1200, 3000

### Related contributions
freifunk-gluon/gluon#2270: Discussion and analysis of several problems
 that were caused by AVM 7430's calibration data being in reversed
 byte order.
#12247: Suggested patch to reverse data in shell script after read out
 by `fritz_cal_extract`.

### Caveats
#### Size constraints
For current devices, `fritz_cal_extract` works on data chunks of a few kB.
This implementation uses a default buffer size presumably large enough
for most common use cases (currently 128kB + 1kB for skipped data).

In most common cases of reading "forward" data, this implementation 
should behave like the original implementation. limit `-l m` will 
determine the amount of data read and size of buffer allocated.

However, if you are reading reversed data or didn't set a limit, the 
default buffer size is used, which may be too small to hold all data. 
In this case, `fritz_cal_extract` will fail, leaving an output file 
of 0 bytes.
You may choose a higher limit `-l m` to enforce a sufficient buffer size.

#### Trailing bytes 
The uncompressed data stream may contain some bytes preceding the 
actual data, which would be skipped in the "forward" case. 
Reversal of data `-r` will put these at the end of the resulting file, 
which may lead to an unexpected file size.

Signed-off-by: Dustin Gathmann <dzsoftware@posteo.org>


Tested-by: Christian Buschau <cbuschau@d00t.de> (@Kistelini)

Kistelini commented:
I tested this, it works. If you want you can add

`Tested-by: Christian Buschau <cbuschau@d00t.de>`

You should get rid of the merge commits and have to add your `Signed-off-by` to the commit message. Furthermore you have to use your real name, see also https://openwrt.org/submitting-patches#submission_guidelines

maurerle commented:
Tested as well, looks good.
the mac-address shows the correct vendor now and there is no wifi 5ghz anymore.

So LGTM!

dzzinstant commented:
Closing this PR in favor of #16352 because of formal flaws.

dzzinstant commented:
I finally managed to rebase this PR branch without breaking the commit history 😔, so this remains the valid PR. Please ignore the duplicate ~~#16352~~ which I announced earlier, and also the mess in the middle of this timeline.

blocktrron commented on package/utils/fritz-tools/src/fritz_cal_extract.c:
```diff
@@ -154,53 +178,59 @@ struct cal_entry {
+			case 's':
+				initial_offset = (int)get_num(optarg);
+				if (errno) {
+					perror("Failed to parse seek offset");
+					goto out_bad;
+				}
+				break;
+			case 'e':
+				entry = (int) htobe16(get_num(optarg));
+				if (errno) {
+					perror("Failed to entry id");
+					goto out_bad;
+				}
+				break;
+			case 'o':
+				out = fopen(optarg, "w");
+				if (!out) {
+					perror("Failed to create output file");
+					goto out_bad;
+				}
+				break;
+			case 'l':
+				limit = (size_t)get_num(optarg);
+				if (errno) {
+					perror("Failed to parse limit");
+					goto out_bad;
+				}
+				break;
+			case 'i':
+				skip = (size_t)get_num(optarg);
+				if (errno) {
+					perror("Failed to parse skip");
+					goto out_bad;
+				}
+				break;
+			case 'r':
+				reversed = true;
+				break;
+			default: /* '?' */
+				usage();
```
Don't reformat the whole switch statement, only add the new case.

dzzinstant replied:
1. Changed commit to match indentation fix [#685e979d470a09361defe1612b001dfc3b7d9832](https://github.com/openwrt/openwrt/commit/685e979d470a09361defe1612b001dfc3b7d9832) from master branch.
2. Also deleted a piece of unreachable code:
```
index 550fd61cd1..c4547bc7e0 100644
--- a/package/utils/fritz-tools/src/fritz_cal_extract.c
+++ b/package/utils/fritz-tools/src/fritz_cal_extract.c
@@ -318,8 +318,6 @@ int main(int argc, char **argv)
 
 	goto out;
 
-	zerr(ret);
-
 out_bad:
 	ret = EXIT_FAILURE;
```

blocktrron reviewed (changes requested):
I've applied the mixed indentation fix to master. In the current state, this can not be applied.

Please don't reformat large blocks of code (See the comment inline). This makes it easier to review.

dzzinstant commented:
1. Changed commit to match indentation fix [#685e979d470a09361defe1612b001dfc3b7d9832](https://github.com/openwrt/openwrt/commit/685e979d470a09361defe1612b001dfc3b7d9832) from master branch.
2. Also deleted a piece of unreachable code:
```
index 550fd61cd1..c4547bc7e0 100644
--- a/package/utils/fritz-tools/src/fritz_cal_extract.c
+++ b/package/utils/fritz-tools/src/fritz_cal_extract.c
@@ -318,8 +318,6 @@ int main(int argc, char **argv)
 
 	goto out;
 
-	zerr(ret);
-
 out_bad:
 	ret = EXIT_FAILURE;
```

maurerle commented:
Hi @dzzinstant, could you rebase this please? :)

maurerle commented:
@dzzinstant there are some formalities in the commit message missing.

@blocktrron - could you take another look at this PR please?

It works for the FB7430, which correctly does not show up with a 5GHz interface anymore and also tested the FB7520 and 7530 with this PR to not break anything.

dzzinstant committed:
lantiq: fritz_cal_extract with reverse option for AVM FritzBox 7430

This implementation of fritz_cal_extract can also retrieve firmware
data stored in reverse byte order, as found in the AVM 7430 device.

This is done by intermediate storage in a buffer presumably large enough
to hold the complete data set. Currently, this buffer size is 128kB + 1kB
(some extra space for skipped data).

In the usual case of "forward" data, this implementation should behave
like the original implementation in all common cases. limit [-l] will
determine the amount of data read and size of buffer allocated.

However, if you are reading reversed data or didn't set a limit, the buffer
may be too small to hold all data. In this case, you can choose a higher
limit [-l] to enforce a sufficient buffer size.

Signed-off-by: Dustin Gathmann <dzsoftware@posteo.org>
Link: https://github.com/openwrt/openwrt/pull/15501
Signed-off-by: Robert Marko <robimarko@gmail.com>

dzzinstant commented:
> @dzzinstant there are some formalities in the commit message missing.

Should be ok now.

maurerle commented:
Currently, the FB7430 does show a wrong MAC address and a broken 5GHz interface which should not be there, due to wrongly extracting the calibration data and mac address from the EEPROM.
This is also documented in https://github.com/freifunk-gluon/gluon/pull/2270.

Both issues are successfully fixed through this PR.

Other devices like the FB7530, which rely on the `fritz_cal_extract` utility as well are working fine with this PR:
https://github.com/openwrt/openwrt/blob/f80f52163a1f56ebf3ced76cb4d37906c4d25702/target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata#L67-L75

Can we get this forward, please? @robimarko @blocktrron

robimarko commented:
@dzzinstant Can you please allow editing by maintainers so I can merge this?

dzzinstant commented:
@robimarko Done.

robimarko commented:
Thanks! Rebased on top of main and merged!

robimarko commented:
Is this required for 24.10 as well?

robimarko commented:
Done, backported
